warp10/src/main/java/io/warp10/script/ListRecursiveStackFunction.java

```diff
@@ -0,0 +1,101 @@
+//
+//   Copyright 2020  SenX S.A.S.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+
+package io.warp10.script;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Abstract class to inherit when a function can be called on a single element or a list of elements or list of list of elements, etc.
+ * This class can be used, for instance, for a function working on GTSs, GTSEncoders and lists thereof, including lists with mixed types.
+ * The idea is to generate a function using the parameters on the stack and then to apply this function on the element
+ * or the list of elements.
+ * This is similar to GTSStackFunction for GTSs but it is a tad faster if the function uses some parameters as this
+ * implementation does not use a Map for parameters.
+ * This is similar to ElementOrListStackFunction but it is faster has it checks for List after the other cases which
+ * it optimal in almost all the cases where the number of elements is higher than the nesting depth of the lists.
+ */
+public abstract class ListRecursiveStackFunction extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  /**
+   * Interface defining the function to be generated which is applied to each element.
+   * This method should check the type of the given element and call unhandled on it if the element is of
```

Comment:
The handling of nested lists should be done by ElementStackFunction itself not left to the willingness of the applyOnElement implementation.

applyOnElement should be declared abstract, but a final apply method should be added which checks if the passed object is a list, and if so it will build the resulting list by calling apply recursively until an element is not a list in which case applyOnElement will be called.

Reply:
After some benchmarks, this solution seems indeed better at the cost of little to none decrease in performance.